pyccel/ast/cudatypes.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/devel/LICENSE for full license details.      #
+#------------------------------------------------------------------------------------------#
+""" Module containing types from the numpy module understood by pyccel
+"""
+from functools import lru_cache
+import numpy as np
+
+from .datatypes import FixedSizeNumericType, HomogeneousContainerType, PythonNativeBool
+from pyccel.utilities.metaclasses import ArgumentSingleton
+from .datatypes import pyccel_type_to_original_type, original_type_to_pyccel_type
+
+
+class CudaArrayType(HomogeneousContainerType, metaclass = ArgumentSingleton):
+    """
+    Class representing the Cuda array type.
+
+    Class representing the Cuda array type
+
+    dtype : NumpyNumericType | PythonNativeBool | GenericType
+        The internal datatype of the object (GenericType is allowed for external
+        libraries, e.g. MPI).
+    rank : int
+        The rank of the new NumPy array.
+    order : str
+        The order of the memory layout for the new NumPy array.
+    memory_location : str
+        The memory location of the new cuda array.
+    """
+    __slots__ = ('_element_type', '_container_rank', '_order', '_memory_location')
+
+    # def __new__(cls, dtype, rank, order, memory_location):
+    #     if rank == 0:
+    #         return dtype
+    #     else:
+    #         return super().__new__(cls, dtype, rank, order)
+    def __init__(self, dtype, rank, order, memory_location):
+        assert isinstance(rank, int)
+        assert order in (None, 'C', 'F')
```

Comment:
Missing assert for memory location

Reply:
Probably also assert rank>0

Reply:
i don't think the rank can be less than 1

Reply:
Neither do I. That's why I want you to add an assert so we have an error if we write code wrong and get something with rank less than 1